README.md

[![Build Status][travis-image]][travis-url]
[![NPM version][npm-version-image]][npm-url]
[![NPM downloads][npm-downloads-image]][npm-url]
[![MIT License][license-image]][license-url]
[![Coverage Status][coverage-image]][coverage-url]
[![Code Quality][codeclimate-image]][codeclimate-url]

# Why

> Oh no, why another parallax script? Do we really need it?

There are many parallax scripts but none of them was satisfying my personal needs:

  - No dependencies
  - No background positioning and heavy obtrusive DOM manipulations
  - Build only for modern devices without internal hacks
  - Modern and flexible api being thought mainly for ajax applications
  - Modern and clean ES6/2015 source code

So I decided to make my own and you can be free to use it or simply ignore it and move forward to the next one!

# Demos

  - [demo](http://gianlucaguarini.github.io/parallax/demo)

# Usage

## Installation

```sh
$ npm install scroll-parallax --save
# or
$ bower install scroll-parallax --save
```

## Markup and initialization

Once you have included the script in your page, you should wrap your parallax elements in a wrapper having an `height`, `position:relative or absolute` and `overflow: hidden`
The elements will be stretched to fit always the whole wrapper size

```html
<figure style="position: relative; height: 300px; overflow: hidden;">
  <img class="parallax" src="path/to/the/image.jpg" />
</figure>

<figure style="position: relative; height: 300px; overflow: hidden;">
  <video class="parallax" src="path/to/the/video.mp4" />
</figure>
```

The Parallax api is really simple and the following snippet should be enough:

```js
var p = new Parallax('.parallax').init()
```

## Options

The options available are only 4 at moment:

| Type    | Name            | Default Value   | Description                                                                                                                                                                         |
|-------- |---------------- |---------------- |------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| Number  | `offsetYBounds` | 50              | the offset top and bottom boundaries in pixels used by the parallax to consider an element in the viewport                                                                            |
| Number  | `intensity`     | 30              | the intensity of the parallax effect                                                                                                                                                |
| Number  | `center`        | 0.5             | the vertical center of the parallax. If you increase this value the element will be centered more on the top of the screen reducing it will look centered at bottom __this value should be between 0 and 1__ |
| Number  | `safeHeight`    | 0.15            | the safe element height gap value in percentage that ensures it can always properly parallax. Any element should be (by default) at least 15% higher than their DOM wrappers (7.5% bottom + 7.5% top) |

You can set the Parallax options in this way:

```js
var p = new Parallax('.parallax', {
  offsetYBounds: 50,
  intensity: 30,
  center: 0.5,
  safeHeight: 0.15
}).init()
```

Each element could be configured using custom Parallax options (except for the `offsetYBounds`) overriding the defaults:

```html
<figure>
  <img class="parallax" data-center="0.8" data-intensity="50" src="path/to/the/image.jpg" />
</figure>
<figure>
  <video class="parallax" data-center="0.2" data-intensity="10" data-safe-height="0.2" src="path/to/the/video.mp4" />
</figure>
```

## API

Each Parallax instance has some useful methods that could be used to adapt it to your application needs

### Parallax.init

Initialize the parallax internal event listeners. The listeners to `element:loaded` and `elements:loaded` should be set before this method gets called

### Parallax.on

The `on` method allows you to listen the internal Parallax events from the outside.<br />
Currently it supports:
  - `element:loaded`: when a parallax elment gets completely loaded
  - `elements:loaded`: when all the elements to parallax get loaded
  - `draw`: when a parallax element becomes visible in the the viewport and starts getting moved
  - `resize`: when the window will be resized
  - `update`: when the page is scrolling and the script has updated all the visible elemens

```js
p.on('element:loaded', function(element){
  // do something with the element
})
```

### Parallax.off

Stop listening an internal Parallax event

```js
var fn = function (element) {
    // do something with the elemnt just drawn
    p.off('draw', fn) // stop listening the draw event
  }
p.on('draw', fn)
```

### Parallax.refresh

Refresh the position of the elements visible in the viewport

```js
// do extremely heavy dom updates
p.refresh()
```

### Parallax.add

Add new element to the parallax instance

```js
// inject new elements
p.add('.parallax-2')
```

### Parallax.remove

Remove element from the parallax instance

```js
p.remove('.parallax-2') // remove the element from the parallax
// and also from the DOM...
```

### Parallax.destroy

Destroy the parallax instance removing all the internal and external callbacks to its internal events

```js
p.destroy() // the parallax is dead!
```

# Contributing

## Available tasks

### Build and test
```shell
$ ./make # or also `$ npm run default`
```

### Convert the ES6 code into valid ES5 combining all the modules into one single file
```shell
$ ./make build # or also `$ npm run build`
```

### Run all the tests
```shell
$ ./make test # or also `$ npm run test`
```

### Start a nodejs static server
```shell
$ ./make serve # or also `$ npm run serve`
```

### To compile and/or test the project anytime a file gets changed
```shell
$ ./make watch # or also `$ npm run watch`
```

[travis-image]:https://img.shields.io/travis/GianlucaGuarini/parallax.svg?style=flat-square
[travis-url]:https://travis-ci.org/GianlucaGuarini/parallax

[license-image]:http://img.shields.io/badge/license-MIT-000000.svg?style=flat-square
[license-url]:LICENSE

[npm-version-image]:http://img.shields.io/npm/v/scroll-parallax.svg?style=flat-square
[npm-downloads-image]:http://img.shields.io/npm/dm/scroll-parallax.svg?style=flat-square
[npm-url]:https://npmjs.org/package/scroll-parallax

[coverage-image]:https://img.shields.io/coveralls/GianlucaGuarini/parallax/master.svg?style=flat-square
[coverage-url]:https://coveralls.io/r/GianlucaGuarini/parallax?branch=master

[codeclimate-image]:https://api.codeclimate.com/v1/badges/05d33fcf100bac2cb67f/maintainability
[codeclimate-url]:https://codeclimate.com/github/GianlucaGuarini/parallax



CHANGELOG.md
## Changelog

### 3.0.3
  * updated: add the `dist` folder to the npm bundle

### 3.0.2
  * updated: reduce the size of the npm package removing demo files

### 3.0.1
  * fixed: refresh method https://github.com/GianlucaGuarini/parallax/issues/16

### 3.0.0
  * added: video capabilities https://github.com/GianlucaGuarini/parallax/pull/19
  * updated: renamed `image:loaded` and `images:loaded` => `element:loaded` and `elements:loaded`
  * fixed: npm vulnerabilites and updated dev dependencies

### 2.1.0
  * added: the possibility to instantiate the `Parallax` class with out DOM selectors
  * updated: improved the destroy method restoring the initial DOM elements state
  * fixed: https://github.com/GianlucaGuarini/parallax/issues/7 thanks to https://github.com/GianlucaGuarini/parallax/pulls/14

### 2.0.2
  * fixed: missing minified version of the script

### 2.0.1
  * added: the `will-change` attribute hoping to gain better performances

### 2.0.0
  * updated: updated riot-observable using the latest version
  * updated: use `transform: matrix()` for better performances
  * fixed: [#5](https://github.com/GianlucaGuarini/parallax/issues/5)

### 1.1.0
  * fixed: https://github.com/GianlucaGuarini/parallax/issues/4 adding the `safeHeight` option

### 1.0.4
  * updated: riot-observable dependency to get faster performances (30% faster)

### 1.0.3
  * fixed: destroy method

### 1.0.2
  * updated: switched from webpack to rollup to get a cleaner dist code

### 1.0.1
  * added: the parallax `update` event
  * added: the `bower.json` file
  * fixed: smoother errors when no images are found

src/index.js
import { $, $$, extend, isUndefined, elementData } from './helpers/helpers'
import Stage from './Stage'
import Canvas from './Canvas'
import o from 'riot-observable'

/**
 * There is no need to listen several times all the window events
 * with this class we listen them once and we subscribe/unsubscribe all the Parallax instances to the main events dispatcher
 * @type {Stage}
 */
let stage

/**
 * Parallax class
 * @param { String } selector - the tag selector where we will bind the parallax
 * @param { Object } opts - parallax options
 */
class Parallax {
  constructor(selector = null, opts = {}) {

    // make this object observable
    o(this)
    // set the options extending the _defaults
    this.opts = opts
    this.selector = selector
    this.canvases = []
    this.bound = false

    // allow to initialize without adding any dom elements
    if (selector !== null) {
      this.add(selector)
    }

    // lazy stage instance initialization
    if (!stage)
      stage = new Stage()

    return this
  }

  /**
   * Initialize the parallax
   * @returns { Object } - Parallax
   */
  init() {
    if (this.bound) {
      throw 'The parallax instance has already been initialized'
    }

    if (!this.canvases.length && this.selector !== null) {
      console.warn(`No elements were found with the selector "${this.selector}"`)
    } else {
      this.elementsLoaded = 0
      this.bind()
    }

    return this
  }

  /**
   * Bind the instance events setting all the callbacks
   * @returns { Object } - Parallax
   */
  bind() {

    // cache these function in order to unbind them when
    // this instance will be destroyed
    this._onResize = (...args) => this.resize.apply(this, args)
    this._onScroll = (...args) => this.scroll.apply(this, args)

    stage.on('resize', this._onResize)
    stage.on('scroll', this._onScroll)

    this.canvases.forEach((canvas) => {
      canvas.one('loaded', () => this.onCanvasLoaded(canvas))
      canvas.load()
    })

    this.bound = true

    return this
  }

  /**
   * Force manually a redraw
   * @returns { Object } - Parallax
   */
  refresh() {
    this._onResize(stage.size)
    this._onScroll(stage.scrollTop)
    return this
  }

  /**
   * Callback triggered once a canvas has loaded its element
   * @param   { Object } canvas - canvas instance
   * @returns { Object } - Parallax
   */
  onCanvasLoaded(canvas) {
    this.trigger('element:loaded', canvas.element, canvas)
    this.elementsLoaded++
    canvas.draw(stage)
    if (this.elementsLoaded == this.canvases.length) this.trigger('elements:loaded')
    return this
  }

  /**
   * Callback triggered on scroll
   * @param   { Number } scrollTop - page offset top
   * @returns { Object } - Parallax
   */
  scroll(scrollTop) {
    const offsetYBounds = this.opts.offsetYBounds,
      { height, width } = stage

    let i = this.canvases.length

    while (i--) {
      let canvas = this.canvases[i],
        canvasHeight = canvas.size.height,
        canvasOffset = canvas.offset

      if (
        canvas.isLoaded &&
          scrollTop + stage.height + offsetYBounds > canvasOffset.top &&
          canvasOffset.top + canvasHeight > scrollTop - offsetYBounds
      ) {
        canvas.draw({ height, scrollTop, width })
        this.trigger('draw', canvas.element)
      }

    }

    this.trigger('update', scrollTop)

    return this
  }

  /**
   * Add parallax elements to this parallax instance
   * @param { String|Array } els - DOM selector or node list
   * @returns { Object } - Parallax
   */
  add(els) {
    const canvases = this.createCanvases($$(els))

    if (this.bound) {
      canvases.forEach((canvas) => {
        canvas.one('loaded', () => this.onCanvasLoaded(canvas))
        canvas.load()
      })
    }

    this.canvases = this.canvases.concat(canvases)

    return this
  }

  /**
   * Remove parallax elements from this parallax instance
   * @param { String|Array } els - DOM selector or node list
   * @returns { Object } - Parallax
   */
  remove(els) {
    $$(els).forEach((el) => {
      let i = this.canvases.length
      while (i--) {
        if (el == this.canvases[i].element) {
          this.canvases[i].destroy()
          this.canvases.splice(i, 1)
          break
        }
      }
    })
    return this
  }

  /**
   * Kill all the internal and external callbacks listening this instance events
   * @returns { Object } - Parallax
   */
  destroy() {
    this.off('*')
    this.canvases = []
    stage.off('resize', this._onResize).off('scroll', this._onScroll)
    return this
  }

  /**
   * Callback triggered on window resize
   * @param   { Object } size - object containing the window width and height
   * @returns { Object } - Parallax
   */
  resize(size) {
    let i = this.canvases.length
    while (i--) {
      const canvas = this.canvases[i]
      if (!canvas.isLoaded) return
      canvas.update().draw(stage)
    }
    this.trigger('resize')
    return this
  }

  /**
   * Set the canvases instances
   * @param   { Array } els - list of the elements we want to parallax
   * @returns { Array } - list of canvas instances
   */
  createCanvases(els) {
    return els.map(el => {
      const data = elementData(el)
      return new Canvas(el, {
        intensity: !isUndefined(data.intensity) ? +data.intensity : this.opts.intensity,
        center: !isUndefined(data.center) ? +data.center : this.opts.center,
        safeHeight: !isUndefined(data.safeHeight) ? +data.safeHeight : this.opts.safeHeight
      })
    })
  }

  /**
   * The options will be always set extending the script _defaults
   * @param   { Object } opts - custom options
   */
  set opts(opts) {
    this._defaults = {
      offsetYBounds: 50,
      intensity: 30,
      center: 0.5,
      // make sure that the elements can always properly parallax
      // They should be at least 15% higher than their wrappers (7.5% bottom + 7.5% top)
      safeHeight: 0.15 // 15%
    }
    extend(this._defaults, opts)
  }

  /**
   * Get the script options object
   * @returns { Object } - current options
   */
  get opts() {
    return this._defaults
  }
}

export default Parallax